### ISSUE TITLE
Navigator is getting recreated after navigation to native UIViewController in iOS

### ISSUE THREAD

**Issue by Shamyyoun** [adrielcafe/voyager#284]
Hi,

I've been using Voyager Navigation library for a couple of weeks in my Compose Multiplatform app.
Here is how I'm initializing it:

```
BottomSheetNavigator(
    modifier = modifier.fillMaxSize(),
    sheetShape = CoreTheme.shapes.sheet,
    sheetElevation = CoreTheme.spacings.sheetElevation,
    sheetBackgroundColor = CoreTheme.colors.sheetBackground,
    skipHalfExpanded = true,
    sheetContent = {
        // Render current bottom sheet inside safe insets container
        BottomSheetInsetsContainer {
            CurrentScreen()
        }
    },
    content = { createdSheetNavigator ->
        // Update current sheet navigator navigator to the created one
        sheetNavigator = createdSheetNavigator

        // Render app navigator
        Navigator(
            screen = startScreen,
            onBackPressed = {
                if (it is BaseScreen<*>) {
                    it.onBackPressed()
                    false
                } else {
                    true
                }
            },
            content = { createdNavigator ->
                // Update current navigator to the created one
                navigator = createdNavigator

                // Render current screen inside safe insets container
                ScreenInsetsContainer(
                    modifier = Modifier.fillMaxSize()
                ) {
                    FadeTransition(createdNavigator)
                }
            }
        )
    }
)
```

And we are required to open the native camera to capture images in the app. But after returning to the app the navigator is recreated and SplashScreen appears again.

And here is how I'm opening the camera to capture images:
```
@Composable
    private fun captureImage() {
        val rootController = LocalUIViewController.current
        imagePickerController.sourceType = UIImagePickerControllerSourceType.UIImagePickerControllerSourceTypeCamera
        rootController.presentViewController(imagePickerController, true) {
            imagePickerController.delegate = delegate
        }
    }
}
```

This only happens in iOS. Android is working fine.

**Comment by ahmedsalemelzeiny** [comment]
@adrielcafe @DevSrSouza any help here?

**Comment by DevSrSouza** [comment]
Can you share how you are displaying the Compose content? I mean, I want the main app Screen that you are adding the ComposeUiViewController to see if I can reproduce here.

If possible, you can share a app containing this issue that I can debug it? Maybe trying to reproduce in the sample app of Voyager and comment here the changes.

**Comment by Shamyyoun** [comment]
Hi @DevSrSouza,

There is no significant or different implementation I'm using. If you have any sample or any running app just add a functionality to open native camera UIViewController for example and then return back to our Compose UIViewController and then you will find it reloaded.

**Comment by Shamyyoun** [comment]
@DevSrSouza 

This is a very quick sample app with the issue,
Just add this file in your iosMain module and render the generated MainViewController in your native iOS app code to see the issue.

Also this is a loom recording with the issue description:
(https://www.loom.com/share/8ce50514742a459396583d347e0d99b7)

Thanks in advance :) 
 

```
package com.example.app

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.interop.LocalUIViewController
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.ComposeUIViewController
import cafe.adriel.voyager.core.screen.Screen
import cafe.adriel.voyager.navigator.LocalNavigator
import cafe.adriel.voyager.navigator.Navigator
import platform.UIKit.UIImagePickerController
import platform.UIKit.UIImagePickerControllerSourceType
import platform.UIKit.UIViewController

/**
 * ================MainViewController================
 */
fun MainViewController() =  ComposeUIViewController {
    // Then render app
    Navigator(
        screen = LandingScreen
    )
}

/**
 * ================Landing Screen================
 */
internal object LandingScreen : Screen {

    @Composable
    override fun Content() {
        val navigator = LocalNavigator.current

        Column(
            modifier = Modifier
                .fillMaxSize()
                .background(Color.White),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(
                space = 2.dp,
                alignment = Alignment.CenterVertically
            )
        ) {
            Text(
                text = "Welcome to Sample App",
                color = Color.Black,
                fontSize = 20.sp
            )

            Text(
                text = "Navigate to Create Profile",
                color = Color.White,
                fontSize = 20.sp,
                modifier = Modifier
                    .background(Color.Blue)
                    .padding(12.dp)
                    .clickable {
                        navigator?.push(
                            CreateProfileScreen
                        )
                    }
            )
        }
    }
}

/**
 * ================Create Profile Screen================
 */
internal object CreateProfileScreen : Screen {

    @Composable
    override fun Content() {
        val navigator = LocalNavigator.current
        val rootViewController = LocalUIViewController.current
        val imagePicker = remember {
            ImagePicker(
                rootController = rootViewController
            )
        }

        Column(
            modifier = Modifier
                .fillMaxSize()
                .background(Color.White),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(
                space = 2.dp,
                alignment = Alignment.CenterVertically
            )
        ) {
            Text(
                text = "Click to Go Back",
                color = Color.White,
                fontSize = 20.sp,
                modifier = Modifier
                    .background(Color.Blue)
                    .padding(12.dp)
                    .clickable {
                        navigator?.pop()
                    }
            )

            Text(
                text = "Open Camera",
                color = Color.White,
                fontSize = 20.sp,
                modifier = Modifier
                    .background(Color.Blue)
                    .padding(12.dp)
                    .clickable {
                        imagePicker.openCamera()
                    }
            )
        }
    }
}

/**
 * ================Image Picker================
 */
internal class ImagePicker(
    private val rootController: UIViewController
) {
    private val pickerController = UIImagePickerController().apply {
        sourceType = UIImagePickerControllerSourceType.UIImagePickerControllerSourceTypeCamera
    }

    fun openCamera() {
        rootController.presentViewController(pickerController, true, null)
    }
}
```

**Comment by DevSrSouza** [comment]
I have being validating your sample, thanks for sharing btw.
It seems that Compose iOS complete destroy the composition when you do a `presentViewController`.

Voyager relies 100% in the `rememberSaveable` API.

Any simple `rememberSaveable` is not being restored in this cenary.
```kotlin
val randomNumber = rememberSaveable {
        Random.nextInt()
    }
Text(text = "$randomNumber")
```

**Comment by DevSrSouza** [comment]
I have open a issue in Compose Multiplatform with your snippet.

https://github.com/JetBrains/compose-multiplatform/issues/4112

**Comment by Shamyyoun** [comment]
@DevSrSouza Thanks Gabriel, hope they fix it soon

**Comment by DevSrSouza** [comment]
I have found a workaround for now.

```kotlin
object ComposeStateRestoration {
    var valuesToBeRestored: MutableState<Map<String, List<Any?>>> = mutableStateOf(emptyMap())
}

fun MainViewController() = ComposeUIViewController {
   CompositionLocalProvider(
        LocalSaveableStateRegistry provides SaveableStateRegistry(
            ComposeStateRestoration.valuesToBeRestored.value,
            canBeSaved = { true }
        )
    ) {
        // ... your app source code here, navigator, etc

       // last item in the code block
       val registry = LocalSaveableStateRegistry.current
       DisposableEffect(Unit) {
            onDispose {
                val savedStates = registry?.performSave()
                ComposeStateRestoration.valuesToBeRestored.value = savedStates ?: emptyMap()
            }
        }
    }
}
```

This solution is only for "single view controller" Compose iOS apps. If you have multiples ViewControllers and use Compose together with UiKit/SwiftUI you can change the `ComposeStateRestoration` to a `class` and add it as a parameter in the function that exposes the ComposeUiViewController (in this example `MainViewController`) and then, hold in the memory a instance of the `ComposeStateRestoration`.

**Comment by DevSrSouza** [comment]
And here is a patch validating the solution on Voyager Multiplatform sample.

```patch
diff --git a/samples/multiplatform/src/iosMain/kotlin/MainViewController.kt b/samples/multiplatform/src/iosMain/kotlin/MainViewController.kt
index dee0b39..48d1463 100644
--- a/samples/multiplatform/src/iosMain/kotlin/MainViewController.kt
+++ b/samples/multiplatform/src/iosMain/kotlin/MainViewController.kt
@@ -1,4 +1,156 @@
+import androidx.compose.foundation.background
+import androidx.compose.foundation.clickable
+import androidx.compose.foundation.layout.Arrangement
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.foundation.layout.padding
+import androidx.compose.material.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.CompositionLocalProvider
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.saveable.LocalSaveableStateRegistry
+import androidx.compose.runtime.saveable.SaveableStateRegistry
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.interop.LocalUIViewController
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
 import androidx.compose.ui.window.ComposeUIViewController
-import cafe.adriel.voyager.sample.multiplatform.SampleApplication
+import cafe.adriel.voyager.core.screen.Screen
+import cafe.adriel.voyager.core.screen.ScreenKey
+import cafe.adriel.voyager.core.screen.uniqueScreenKey
+import cafe.adriel.voyager.navigator.LocalNavigator
+import cafe.adriel.voyager.navigator.Navigator
+import platform.UIKit.UIImagePickerController
+import platform.UIKit.UIImagePickerControllerSourceType
+import platform.UIKit.UIViewController
 
-fun MainViewController() = ComposeUIViewController { SampleApplication() }
+object ComposeStateRestoration {
+    var valuesToBeRestored: MutableState<Map<String, List<Any?>>> = mutableStateOf(emptyMap())
+}
+
+fun MainViewController() = ComposeUIViewController {
+    CompositionLocalProvider(
+        LocalSaveableStateRegistry provides SaveableStateRegistry(
+            ComposeStateRestoration.valuesToBeRestored.value,
+            canBeSaved = { true }
+        )
+    ) {
+        val registry = LocalSaveableStateRegistry.current
+
+        Navigator(LandingScreen)
+
+        DisposableEffect(Unit) {
+            onDispose {
+                val savedStates = registry?.performSave()
+                ComposeStateRestoration.valuesToBeRestored.value = savedStates ?: emptyMap()
+            }
+        }
+    }
+}
+
+internal data object LandingScreen : Screen {
+    override val key: ScreenKey = uniqueScreenKey
+
+    @Composable
+    override fun Content() {
+        val navigator = LocalNavigator.current
+
+        Column(
+            modifier = Modifier
+                .fillMaxSize()
+                .background(Color.White),
+            horizontalAlignment = Alignment.CenterHorizontally,
+            verticalArrangement = Arrangement.spacedBy(
+                space = 2.dp,
+                alignment = Alignment.CenterVertically
+            )
+        ) {
+            Text(
+                text = "Welcome to Sample App",
+                color = Color.Black,
+                fontSize = 20.sp
+            )
+
+            Text(
+                text = "Navigate to Create Profile",
+                color = Color.White,
+                fontSize = 20.sp,
+                modifier = Modifier
+                    .background(Color.Blue)
+                    .padding(12.dp)
+                    .clickable {
+                        navigator?.push(
+                            CreateProfileScreen
+                        )
+                    }
+            )
+        }
+    }
+}
+
+internal data object CreateProfileScreen : Screen {
+    override val key: ScreenKey = uniqueScreenKey
+
+    @Composable
+    override fun Content() {
+        val navigator = LocalNavigator.current
+        val rootViewController = LocalUIViewController.current
+        val imagePicker = remember {
+            ImagePicker(
+                rootController = rootViewController
+            )
+        }
+
+        Column(
+            modifier = Modifier
+                .fillMaxSize()
+                .background(Color.White),
+            horizontalAlignment = Alignment.CenterHorizontally,
+            verticalArrangement = Arrangement.spacedBy(
+                space = 2.dp,
+                alignment = Alignment.CenterVertically
+            )
+        ) {
+            Text(
+                text = "Click to Go Back",
+                color = Color.White,
+                fontSize = 20.sp,
+                modifier = Modifier
+                    .background(Color.Blue)
+                    .padding(12.dp)
+                    .clickable {
+                        navigator?.pop()
+                    }
+            )
+
+            Text(
+                text = "Open Camera",
+                color = Color.White,
+                fontSize = 20.sp,
+                modifier = Modifier
+                    .background(Color.Blue)
+                    .padding(12.dp)
+                    .clickable {
+                        imagePicker.openCamera()
+                    }
+            )
+        }
+    }
+}
+
+internal class ImagePicker(
+    private val rootController: UIViewController
+) {
+    private val pickerController = UIImagePickerController().apply {
+        sourceType = UIImagePickerControllerSourceType.UIImagePickerControllerSourceTypeCamera
+    }
+
+    fun openCamera() {
+        rootController.presentViewController(pickerController, true, null)
+    }
+}
```

Steps the apply the code:
- Create a file `state_restauration_workaround.patch` with the above patch
- run `git apply state_restauration_workaround.patch`

**Comment by DevSrSouza** [comment]
This is not a Voyager issue, this workaround will do the work.
I will keep this issue open until we have proper documentation.

**Comment by Shamyyoun** [comment]
Hi @DevSrSouza,
I tried your workaround but it's not working also.

But I found a brilliant solution by luck while looking for some library and is working fine for me


```
//
//  AvoidDisposeViewController.swift
//  iosApp
//
//  Created by Clem on 27.10.22.
//  Copyright © 2022 orgName. All rights reserved.
//

import UIKit
import SwiftUI

struct AvoidDisposeViewControllerRepresentable: UIViewControllerRepresentable {
    let composeController: UIViewController
    
    func makeUIViewController(context: Context) -> UIViewController {
        return AvoidDisposeViewController(composeController)
    }
    
    func updateUIViewController(_ uiViewController: UIViewController, context: Context) {}
}


class AvoidDisposeViewController: UIViewController {
    let controller: UIViewController
    
    init(_ controller: UIViewController) {
        self.controller = controller
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        addChild(controller)
        view.addSubview(controller.view)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        controller.removeFromParent()
        controller.view.removeFromSuperview()
    }
    
    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)
        skiaRefresh()
    }
    
    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        super.traitCollectionDidChange(previousTraitCollection)
        skiaRefresh()
    }
    
    override func viewSafeAreaInsetsDidChange() {
        super.viewSafeAreaInsetsDidChange()
        skiaRefresh()
    }
    
    private func skiaRefresh() {
        controller.view.frame = view.bounds
        controller.viewWillAppear(false)
        controller.view.touchesCancelled([UITouch()], with: UIEvent())
    }
}
```

**Comment by DevSrSouza** [comment]
I think it should be fixed on Compose KMP [1.6.10-beta01](https://github.com/JetBrains/compose-multiplatform/releases/tag/v1.6.10-beta01)

**Comment by Shamyyoun** [comment]
Tested this issue gain on compose multiplatform 1.6.11 and it's working fine 🥳🥳
